### ISSUE TITLE
Updates of learnable scales following L2 normalization randomly change for constant inputs

### ISSUE THREAD

**Issue by PuchatekwSzortach** [keras-team/keras#7183]
Hello

I'm trying to test a custom normalization layer (for Single Shot Detector implementation) but in my tests I'm seeing that after calling `model.train_on_batch(x, y)` about 1/4th the time I get expected results, but 3/4th of the time I get different values. Maybe someone can point out why is that.

Here's the layer definition:
```python
class Normalize(keras.layers.Layer):

    def __init__(self, scale, **kwargs):
	
        self.channel_axis = 3
        self.scale = scale
        super(Normalize, self).__init__(**kwargs)
	
    def build(self, input_shape):
	
        self.input_spec = [keras.engine.topology.InputSpec(shape=input_shape)]
	
        # Create scaling gammas, one for each channel
        self.gamma = self.add_weight(
            name='{}_gamma'.format(self.name),
            shape=(input_shape[self.channel_axis],),
            initializer=keras.initializers.Constant(self.scale),
            trainable=True)
	
        super(Normalize, self).build(input_shape)
	
    def call(self, x, mask=None):
	
        # Normalize along channel axis
        output = keras.backend.l2_normalize(x, self.channel_axis)
	
        # Scale by gamma and return
        return output * self.gamma
```

And here is my test:

```python
def test_normalize_training_input_has_channels_only():

    keras.backend.clear_session()

    def get_model():

        input = keras.layers.Input(shape=(1, 1, 2))

        x = face.layers.Normalize(scale=1, name='normalization_layer')(input)
        x = keras.layers.Flatten()(x)

        initializer = keras.initializers.Constant(1)
        x = keras.layers.Dense(units=1, kernel_initializer=initializer, name='dense_layer')(x)

        model = keras.models.Model(inputs=input, outputs=x)

        optimizer = keras.optimizers.SGD(lr=1)
        model.compile(optimizer=optimizer, loss='mse')

        return model

    model = get_model()

    # Test initial values of normalization layer scales and dense layer weights
    # are as expected
    normalization_layer = model.layers[1]

    expected_initial_normalization_scales = np.array([1, 1])
    actual_initial_normalization_scales = keras.backend.eval(normalization_layer.scales)

    assert np.all(expected_initial_normalization_scales == actual_initial_normalization_scales)

    dense_layer = model.layers[3]
    expected_initial_weights = np.array([1, 1]).reshape(2, 1)
    actual_initial_weights = keras.backend.eval(dense_layer.kernel)

    assert np.all(expected_initial_weights == actual_initial_weights)

    x = np.array([2, 4]).reshape(1, 1, 1, 2)
    y = np.array([10])

    expected_initial_bias = np.array([0])
    actual_initial_bias = keras.backend.eval(dense_layer.bias)

    assert np.all(expected_initial_bias == actual_initial_bias)

    expected_prediction = np.array(6 / np.sqrt(20))
    actual_prediction = model.predict(x)

    assert np.allclose(expected_prediction, actual_prediction)

    expected_loss = np.array(np.power([10 - (6 / np.sqrt(20))], 2))
    actual_loss = model.train_on_batch(x, y)

    assert np.allclose(expected_loss, actual_loss)

    expected_weights = np.array([8.7442, 16.4885]).reshape(2, 1)
    actual_weights = keras.backend.eval(dense_layer.kernel)

    assert np.allclose(expected_weights, actual_weights, atol=0.01)

    expected_bias = np.array([17.3168])
    actual_bias = keras.backend.eval(dense_layer.bias)

    assert np.allclose(expected_bias, actual_bias)

    expected_scales = np.array([8.7442, 16.4885])
    actual_scales = keras.backend.eval(normalization_layer.scales)

    # This assertion randomly fails,
    # about ~75% of time actual_scales are [68.71802521,  256.38354492],
    # while ~25% of time actual_scales equal expected_scales
    assert np.allclose(expected_scales, actual_scales, atol=0.01)
```

**Comment by kuba-lexues** [comment]
For comparison here's the same code implemented in pure Tensorflow, along with a test that always passes:

```python
input = tf.placeholder(dtype=tf.float32, shape=[1, 2])

normalized = tf.nn.l2_normalize(input, dim=-1)
scales = tf.Variable(np.array([1.0, 1.0]).reshape(1, 2), dtype=tf.float32)

scaled_normalized_output = scales * normalized

weights = tf.Variable(np.array([1, 1]).reshape(2, 1), dtype=tf.float32)
output = tf.matmul(scaled_normalized_output, weights)

true_output = tf.placeholder(dtype=tf.float32, shape=[1])
loss = tf.square(true_output - output)

train_op = tf.train.GradientDescentOptimizer(learning_rate=1.0).minimize(loss)

x = np.array([2, 4]).reshape(1, 2)
y = np.array([10])

with tf.Session() as session:

    session.run(tf.global_variables_initializer())

    feed_dictionary = {input: x}

    expected_scaled_normalized_output = np.array([2 / np.sqrt(20), 4 / np.sqrt(20)]).reshape(1, 2)
    actual_scaled_normalized_output = session.run(scaled_normalized_output, feed_dictionary)
    assert np.allclose(expected_scaled_normalized_output, actual_scaled_normalized_output)

    expected_output = np.array([6 / np.sqrt(20)]).reshape(1, 1)
    actual_output = session.run(output, feed_dictionary)
    assert np.isclose(expected_output, actual_output)

    feed_dictionary = {input: x, true_output: y}

    expected_loss = np.array(np.square(10 - (6 / np.sqrt(20))))
    actual_loss = session.run(loss, feed_dictionary)
    assert np.allclose(expected_loss, actual_loss, atol=0.001)

    _, actual_weights, actual_scales = session.run([train_op, weights, scales], feed_dictionary)

    expected_weights = np.array([8.744, 16.4888]).reshape(2, 1)
    assert np.allclose(expected_weights, actual_weights, atol=0.001)

    expected_scales = np.array([8.744, 16.4888]).reshape(1, 2)
    assert np.allclose(expected_scales, actual_scales, atol=0.001)
```

**Comment by PuchatekwSzortach** [comment]
For comparison here's the same logic implemented in pure Tensorflow, along with a test that always passes:

```python
input = tf.placeholder(dtype=tf.float32, shape=[1, 2])

normalized = tf.nn.l2_normalize(input, dim=-1)
scales = tf.Variable(np.array([1.0, 1.0]).reshape(1, 2), dtype=tf.float32)

scaled_normalized_output = scales * normalized

weights = tf.Variable(np.array([1, 1]).reshape(2, 1), dtype=tf.float32)
output = tf.matmul(scaled_normalized_output, weights)

true_output = tf.placeholder(dtype=tf.float32, shape=[1])
loss = tf.square(true_output - output)

train_op = tf.train.GradientDescentOptimizer(learning_rate=1.0).minimize(loss)

x = np.array([2, 4]).reshape(1, 2)
y = np.array([10])

with tf.Session() as session:

    session.run(tf.global_variables_initializer())

    feed_dictionary = {input: x}

    expected_scaled_normalized_output = np.array([2 / np.sqrt(20), 4 / np.sqrt(20)]).reshape(1, 2)
    actual_scaled_normalized_output = session.run(scaled_normalized_output, feed_dictionary)
    assert np.allclose(expected_scaled_normalized_output, actual_scaled_normalized_output)

    expected_output = np.array([6 / np.sqrt(20)]).reshape(1, 1)
    actual_output = session.run(output, feed_dictionary)
    assert np.isclose(expected_output, actual_output)

    feed_dictionary = {input: x, true_output: y}

    expected_loss = np.array(np.square(10 - (6 / np.sqrt(20))))
    actual_loss = session.run(loss, feed_dictionary)
    assert np.allclose(expected_loss, actual_loss, atol=0.001)

    _, actual_weights, actual_scales = session.run([train_op, weights, scales], feed_dictionary)

    expected_weights = np.array([8.744, 16.4888]).reshape(2, 1)
    assert np.allclose(expected_weights, actual_weights, atol=0.001)

    expected_scales = np.array([8.744, 16.4888]).reshape(1, 2)
    assert np.allclose(expected_scales, actual_scales, atol=0.001)
```

**Comment by PuchatekwSzortach** [comment]
There is a bug in the code that's on Keras side and I provided tests to prove that.

**Comment by fabriciorsf** [comment]
Sorry, but the value of self.channel_axis should be 2 (instead 3), shouldn't?
```
output = keras.backend.l2_normalize(x, self.channel_axis)
```

**Comment by kuba-lilz** [comment]
@fabriciorsf Why? First axis (index 0) is for batch, second (index 1) for y-dimension, third (index 2) for x-dimension and fourth (index 3) for channel.

**Comment by fabriciorsf** [comment]
Sorry, for a moment I forgot the first axis is for batch.
Do you know if is it valid to metric functions inputs too?

**Comment by PuchatekwSzortach** [comment]
Not sure what do you mean by "metric functions inputs" or how does that relate to this issue.

**Comment by rdelassus** [comment]
Something to do with that stack overflow question maybe ? 
https://stackoverflow.com/questions/41064802/l2-normalised-output-with-keras

TL;DR tensorflow has an optional argument "epsilon". "If the output of the net contains numbers lower than epsilon (which is set to 1e-12 by default) then it is not normalized correctly"

**Comment by PuchatekwSzortach** [comment]
@rdelassus As you can see in tests, net outputs values between 0.1 and 10 at its different stages.
